torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -0,0 +1,307 @@
+            if (II != I->second.end())
+              paramConst = II->second;
+          }
+          if (!paramConst) {
+            auto attr = attrModule.attr(name);
+            if (!shouldFoldAttr(attr, name, isEval))
+              continue;
+            if (auto attrVal = tryInsertConstant(*graph, attr)) {
+              paramConst = *attrVal;
+            } else {
+              GRAPH_DEBUG(
+                  attr.type()->cast<ClassType>() ? "" : "attribute: ",
+                  name,
+                  " is not materializable.");
+              continue;
+            }
+            auto moduleName = attrModule.type()->name()->qualifiedName();
+            moduleName += ".";
+            moduleName += name;
+            paramConst->setDebugName(moduleName);
+            attrValues[attrModule._ivalue()][name] = paramConst;
+          }
+          GRAPH_UPDATE(
+              "Folding GetAttr %",
+              n->outputs()[0]->debugName(),
+              " with ",
+              paramConst->debugName());
+          n->outputs().at(0)->replaceAllUsesWith(paramConst);
+          n->removeAllInputs();
+        }
+      }
+    }
+  }
+
+  // cleanupFrozenModule function cleans up the Frozen module. it performs the
+  // following:
+  // 1) Remove unused attributes.
+  // 2) Remove unreferenced submodules
+  // 3) Remove non pulic unreferenced methods.
+  // TODO: do #3 because there is no API to 'unsafely' remove methods.
+  void cleanupFrozenModule(std::shared_ptr<Graph>& graph) {
+    std::vector<std::string> attrsToRemove;
+    auto type = module_.type();
+    size_t N = type->numAttributes();
+    auto KeepAttrs = recordReferencedAttrs(graph);
+    for (size_t i = 0; i < N; ++i) {
+      auto attrTy = type->getAttribute(i);
+      auto name = type->getAttributeName(i);
+      if (!KeepAttrs.count(name)) {
+        attrsToRemove.push_back(name);
+      }
+    }
+    for (auto& name : attrsToRemove) {
+      module_._ivalue()->unsafeRemoveAttr(name);
+      module_.type()->unsafeRemoveAttribute(name);
+    }
+  }
+
+  // Contains attributes that can't be folded or user directs to keep them.
+  std::unordered_map<script::ModulePtr, std::set<std::string>> preservedAttrs_;
+
+  script::Module& module_;
+}; // class AttributePropagator
+} // namespace
+
+script::Module freeze_module(const script::Module& module) {
+  // Currently freezing module is supported only in eval mode.
+  // If assertion below is commented and module is in training mode then this
+  // implementation folds attributes correctly. Tensor attributes with
+  // required_grad set are not folded and 'training' attribute is also not
+  // folded.
+  // TODO: Determine if freezing in training mode is useful and further clarify
+  // its semantics.
+  TORCH_INTERNAL_ASSERT(!module.is_training());
+  auto moduleClone = module.clone();
+  AttributePropagator attrPropagator(moduleClone);
+  script::Method method = moduleClone.get_method("forward");
```

Comment:
Noting from offline discussion: if we are exposing more than just `forward()` of the frozen module, we need to do interprocedural analysis to make sure that some method (e.g. `mutate_foo()`) is not mutating an attr that got folded by another (`use_foo()`).

Reply:
Correct! the interface is a follow up patch :)